Document Snaps data storage (#1278)
MetaMask/metamask-docs: snaps/features/data-storage.md

sidebar_position: 5
---

# Data storage

You can store and manage sensitive information within a Snap using encrypted storage, or
non-sensitive information using unencrypted storage.
Use the [`snap_manageState`](../reference/snaps-api.md#snap_managestate) API method to persist up to
100 MB of data to the user's disk and retrieve it at will.
We recommend using this method for storing data in a Snap long term.

## Steps

### 1. Get permission to store data

Request the [`snap_manageState`](../reference/snaps-api.md#snap_managestate) permission.
Add the following to your Snap's manifest file:

```json title="snap.manifest.json"
"initialPermissions": {
  "snap_manageState": {}
}
```

### 2. Use encrypted storage

By default, [`snap_manageState`](../reference/snaps-api.md#snap_managestate) automatically encrypts
data using a Snap-specific key before storing it on the user's disk, and automatically decrypts it
when retrieved.
This is useful to store sensitive information, such as passwords.

The following example uses `snap_manageState` to store some data using the `update` operation, and
retrieves the data at a later time using the `get` operation.

> Once there's no need to store the data anymore, it clears the Snap's state using the `clear` operation.

Reviewer: I think refrain from using "once" as a synonym for after 
https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/o/once

```suggestion
When the data is no longer required, the Snap's state is cleared using the `clear` operation.
```

```javascript title="index.js"
// Persist some data.
await snap.request({
  method: "snap_manageState",
  params: { 
    operation: "update",
    newState: { hello: "world" },
  },
});

// At a later time, get the stored data.
const persistedData = await snap.request({
  method: "snap_manageState",
  params: { operation: "get" },
});

console.log(persistedData);
// { hello: "world" }

> // If there's no need to store data anymore, clear it out.

Reviewer: ```suggestion
// // If data storage is no longer necessary, clear it.
```

[`snap_getClientStatus`](../reference/snaps-api.md#snap_getclientstatus) to ensure that MetaMask is
unlocked before accessing state, preventing an unexpected password request.
:::

### 3. Use unencrypted storage

To use unencrypted storage, set `encrypted` to `false` when storing, retrieving, or clearing data
using [`snap_manageState`](../reference/snaps-api.md#snap_managestate).
The Snap will use a storage section separate from the encrypted storage, and will not encrypt the data.
This is useful to access non-sensitive data from background operations such as
[cron jobs](cron-jobs.md), without requiring the user to enter their password in the case that
MetaMask is locked.

```javascript title="index.js"
// Persist some data.
await snap.request({
  method: "snap_manageState",
  params: { 
    operation: "update",
    newState: { hello: "world" },
    encrypted: false,
  },
});

// At a later time, get the stored data.
const persistedData = await snap.request({
  method: "snap_manageState",
  params: { 
    operation: "get",
    encrypted: false,
  },
});

console.log(persistedData);
// { hello: "world" }

> // If there's no need to store data anymore, clear it out.

Reviewer: ```suggestion
// If data storage is no longer necessary, clear it.
```
